object_tracker/match.py

```diff
@@ -0,0 +1,151 @@
+        matches = sorted(matches, key=lambda x: x.distance)
+        matches = matches[: self._params.num_orb_features]
+
+        list_kp1 = [kp1[mat.queryIdx].pt for mat in matches]
+        list_kp2 = [kp2[mat.trainIdx].pt for mat in matches]
+        if len(list_kp1) < self._params.min_points_for_find_homography:
+            return None, None
+        M, mask = cv2.findHomography(
+            np.squeeze(list_kp1), np.squeeze(list_kp2), cv2.RANSAC, 5.0
+        )
+        return M, mask
+
+    def _calc_optical_flow(self, p0, image):
+        """
+        this method calculates a homography matrix using OpticalFlowPyrLK
+        return M(homography matrix), mask
+        """
+        p1, st1, err1 = cv2.calcOpticalFlowPyrLK(
+            self._prev_image, image, p0, None, **self._params.tracking_params
+        )
+        if self._params.bidirectional_enable:
+            p2, st2, err2 = cv2.calcOpticalFlowPyrLK(
+                image, self._prev_image, p1, None, **self._params.tracking_params
+            )
+            proj_err = np.linalg.norm(p2 - p0, axis=1)
+            st = np.squeeze(st1 * st2) * (
+                proj_err < self._params.bidirectional_thresh
+            ).astype("uint8")
+        else:
+            st = np.squeeze(st1)
+        if len(np.squeeze(p0[st > 0])) < self._params.min_points_for_find_homography:
+            logging.error("failed to match features with optical flow")
+            return None, None
+        M, mask = cv2.findHomography(
+            np.squeeze(p0[st > 0]), np.squeeze(p1[st > 0]), cv2.RANSAC, 5.0
+        )
+        return M, mask
+
+    def _calc_new_roi(self, M):
+        """
+        this method calculates a new roi using the M(homography matrix).
+        return: min_x, min_y, max_x, max_y
+        """
+        top_left = [self.roi.min_x, self.roi.min_y]
+        top_right = [self.roi.min_x + self.roi.w, self.roi.min_y]
+        bottom_right = [self.roi.min_x + self.roi.w, self.roi.min_y + self.roi.h]
+        bottom_left = [self.roi.min_x, self.roi.min_y + self.roi.h]
+        coord_of_roi = np.array([top_left, top_right, bottom_right, bottom_left]).reshape(-1, 1, 2)
+        transform_points = cv2.perspectiveTransform(
+            coord_of_roi.astype(np.float), M
+        ).reshape(-1, 2)
+        min_x = np.min(transform_points, axis=0)[0]
+        min_y = np.min(transform_points, axis=0)[1]
+        max_x = np.max(transform_points, axis=0)[0]
+        max_y = np.max(transform_points, axis=0)[1]
+        w = max_x - min_x
+        h = max_y - min_y
+        self.roi = Roi(min_x=min_x, min_y=min_y, w=w, h=h)
+        return min_x, min_y, max_x, max_y
+
+    def __call__(self, image):
+        """
+        this method runs optical flow and orb(if necessary) on the image
+        return: min_x, min_y, max_x, max_y
+        """
+        image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
+        p0 = np.squeeze(self.features)
+
+        if len(p0) == 0 or len(p0.shape) < 2:
```

Comment:
Document this if meaning (no enough features)

Reply:
ok

Reply:
done